sdk/python/lib/pulumi/resource.py

```diff
@@ -796,26 +795,23 @@ def __init__(self,
             # Infer providers and provider maps from parent, if one was provided.
             self._providers = opts.parent._providers
 
-        if custom:
-            provider = opts.provider
-            if provider is None:
-                if not opts.parent is None:
-                    # If no provider was given, but we have a parent, then inherit the
-                    # provider from our parent.
-                    opts.provider = opts.parent.get_provider(t)
-            else:
-                # If a provider was specified, add it to the providers map under this type's package
-                # so that any children of this resource inherit its provider.
-                type_components = t.split(":")
-                if len(type_components) == 3:
-                    [pkg, _, _] = type_components
-                    self._providers = {**self._providers, pkg: provider}
-        else:
-            providers = convert_providers(opts.provider, opts.providers)
-            self._providers = {**self._providers, **providers}
+        type_components = t.split(":")
+        pkg = None
+        if len(type_components) == 3:
+            [pkg, _, _] = type_components
+
+        opts.provider, providers = self._get_providers(t, pkg, opts)
 
         self._protect = bool(opts.protect)
         self._provider = opts.provider if custom else None
+        if self._provider and self._provider.package != pkg:
+            action = (
+                (opts.urn is not None and "get") or
```

Comment:
Oof.

`"get" if opts_run is None else "read" if opts_id is None else "register"`?